File: libs/shared/src/types/message.ts
import type { BaseRecord, CursorPaginatedResponse } from './base';
import type { ReactionSummary } from './reaction';

/**
 * Message record from messages table
 */
export interface MessageRecord extends BaseRecord {
  content: string;
  channel: string;
  sender: string;
  type: 'text' | 'system' | 'giphy' | 'image';
  gif_url?: string;
  image_url?: string;
  image_width?: number | null;
  image_height?: number | null;
  seq: number;
}

export type Message = Pick<
  MessageRecord,
  | 'id'
  | 'content'
  | 'channel'
  | 'sender'
  | 'type'
  | 'gif_url'
  | 'image_url'
  | 'image_width'
  | 'image_height'
  | 'seq'
  | 'created_at'
  | 'updated_at'
>;

/**
 * Message with expanded sender data
 */
export interface MessageWithSender {
  id: string;
  content: string;
  channel: string;
  sender: {
    id: string;
    username: string;
    avatar?: string;
  };
  type: 'text' | 'system' | 'giphy' | 'image';
  gif_url?: string;
  image_url?: string;
  image_width?: number | null;
  image_height?: number | null;
  seq: number;
  created_at: string;
  updated_at: string;
  reactions?: ReactionSummary[];
}

export interface MessageListResponse {
  messages: MessageWithSender[];
  totalPages: number;
  page: number;
}

export type CursorPaginatedMessages =
  CursorPaginatedResponse<MessageWithSender>;

export interface MessageSearchRequest {
  query: string;
  channelId?: string;
  groupId: string;
  cursor?: string;
  limit?: number;
}

export type MessageSearchResponse = CursorPaginatedResponse<MessageWithSender>;

File: apps/chat-app-dj/chat-app-dj/Models/Message.swift
import Foundation

// MARK: - Message Type

enum MessageType: String, Codable, Sendable {
    case text
    case system
    case giphy
    case image
}

// MARK: - Message Sender

nonisolated struct MessageSender: Codable, Equatable, Sendable {
    let id: String
    let username: String
    let avatar: String?
}

// MARK: - Message with Sender

nonisolated struct MessageWithSender: Codable, Identifiable, Equatable, Sendable {
    let id: String
    let seq: Int
    let content: String
    let channel: String
    let sender: MessageSender
    let type: MessageType
    let gifUrl: String?
    let imageUrl: String?
    let imageWidth: Int?
    let imageHeight: Int?
    let createdAt: String
    let updatedAt: String
    var reactions: [ReactionSummary]?

    enum CodingKeys: String, CodingKey {
        case id, seq, content, channel, sender, type, reactions
        case gifUrl = "gif_url"
        case imageUrl = "image_url"
        case imageWidth = "image_width"
        case imageHeight = "image_height"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}
